src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameProjection.java

```diff
@@ -0,0 +1,461 @@
+            startingSequenceNumber, sequenceNumberDelta);
+        pkt.setSequenceNumber(sequenceNumber);
+
+        byte[] buf = pkt.getBuffer();
+        int payloadOff = pkt.getPayloadOffset()
+            , payloadLen = pkt.getPayloadLength();
+
+        if (!DePacketizer.VP8PayloadDescriptor.setTL0PICIDX(
+            buf, payloadOff, payloadLen, tl0PICIDX))
+        {
+            logger.warn("Failed to set the TL0PICIDX of a VP8 packet.");
+        }
+
+        if (!DePacketizer.VP8PayloadDescriptor.setExtendedPictureId(
+            buf, payloadOff, payloadLen, extendedPictureId))
+        {
+            logger.warn("Failed to set the picture id of a VP8 packet.");
+        }
+    }
+
+    /**
+     * Determines whether a packet can be forwarded as part of this
+     * {@link VP8FrameProjection} instance. The check is based on the sequence
+     * of the incoming packet and whether or not the {@link VP8FrameProjection}
+     * instance is the "last" {@link VP8FrameProjection} or not.
+     *
+     * @param rtpPacket the {@link RawPacket} that will be examined .
+     * @return true if the packet can be forwarded as part of this
+     * {@link VP8FrameProjection}, false otherwise.
+     */
+    public boolean accept(@NotNull RawPacket rtpPacket)
+    {
+        if (vp8Frame == null || !vp8Frame.matchesFrame(rtpPacket))
+        {
+            // The packet does not belong to this VP8 picture.
+            return false;
+        }
+
+        synchronized (vp8Frame)
+        {
+            int sequenceNumber = rtpPacket.getSequenceNumber();
+            int deltaFromMax = RTPUtils.getSequenceNumberDelta(
+                vp8Frame.getMaxSequenceNumber(), sequenceNumber);
+
+            boolean isGreaterThanMax
+                = vp8Frame.getMaxSequenceNumber() == -1 || deltaFromMax < 0;
+
+            if (isLast)
+            {
+                if (isGreaterThanMax)
+                {
+                    vp8Frame.setMaxSequenceNumber(
+                        sequenceNumber, rtpPacket.isPacketMarked());
+                }
+
+                return true;
+            }
+            else
+            {
+                return !isGreaterThanMax;
+            }
+        }
+    }
+
+    /**
+     * @return The projected {@link VP8Frame}.
+     */
+    VP8Frame getVP8Frame()
+    {
+        return vp8Frame;
+    }
+
+    /**
+     * @return The RTP SSRC of this projection.
+     */
+    public long getSSRC()
+    {
+        return ssrc;
+    }
+
+    /**
+     * @return true if we have projected all the packets from starting sequence
+     * number to max sequence number.
+     */
+    boolean isFullyProjected(long nowMs)
+    {
+        // Assume a frame is fully projected after WAIT_MS. A more predictable
+        // approach would be to check if all the packets have been projected,
+        // i.e. keep a bitmap of received packets of a frame.
+        return nowMs - createdMs > WAIT_MS;
```

Comment:
i'm surprised to find this method return a value based on time, but maybe it's because i don't understand the concept of 'projection' fully.  but your comment does make sense to me (that it would be based on having all the packets of a frame) but i could also see some concept of a 'timeout' being useful (as i see this is used for cleaning out the map)

Reply:
The idea is that after we've fully projected all the packets of a frame, we no longer need the frame projection instance and it can be expired/discarded to prevent the map that stores these projections from growing too big. Any re-transmissions of that frame can be safely dropped and, basically, the frame projection instance becomes completely useless.

To implement this correctly we need 1) a check that determines if all the packets of the projected frame have been projected and 2nd) an upper bound of the time we're willing to wait for the packets to arrive and be projected (because the sender may decide to never send some packets).

Due to lack of time in this method I've only implemented the 2nd, easier part. Does that make more sense?

Reply:
> Due to lack of time in this method I've only implemented the 2nd, easier part

Lack of time?  We _do_ have time here, don't we?  Did you mean a lack of packet knowledge?

Reply:
oh, I mean I have personally have run out of bandwidth :)